### PR TITLE
Option Chain Provider Symbol Map Resolve

### PR DESCRIPTION
**Pull request by C-SELLERS**
#### Description

Have the backtesting Option chain provider resolve and check the underlying symbols map file for different tickers to ensure that it is able to fetch data from the appropriate location

#### Related Issue

Related #5636 

#### Motivation and Context

This issue is most likely to occur in research but also in algorithm history requests that call on GetOptionContracts.

Basically if the symbol is not resolved the data key the option chain provider creates may be incorrect, leading to it failing to retrieve the data.

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

Tested locally with research and algorithms
Added reproducing unit tests (red/green from master/this branch)

Running regressions currently

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Engine/DataFeeds/BacktestingOptionChainProvider.cs [hunk 1]
```diff
@@ -51,13 +56,19 @@ public IEnumerable<Symbol> GetOptionContractList(Symbol underlyingSymbol, DateTi
                 throw new NotSupportedException($"BacktestingOptionChainProvider.GetOptionContractList(): SecurityType.Equity, SecurityType.Future, or SecurityType.Index is expected but was {underlyingSymbol.SecurityType}");
             }
 
+            // Resolve any mapping before requesting option contract list
+            var mapFileResolver = _mapFileProvider.Get(underlyingSymbol.ID.Market);
+            var mapFile = mapFileResolver.ResolveMapFile(underlyingSymbol.Value, date);
```
Comment (Martin-Molinero) [inline Engine/DataFeeds/BacktestingOptionChainProvider.cs:61]: Think should send in `.ID.Symbol` here, can see other users of `.ResolveMapFile()`

File: Engine/DataFeeds/BacktestingOptionChainProvider.cs [hunk 2]
```diff
@@ -51,13 +56,19 @@ public IEnumerable<Symbol> GetOptionContractList(Symbol underlyingSymbol, DateTi
                 throw new NotSupportedException($"BacktestingOptionChainProvider.GetOptionContractList(): SecurityType.Equity, SecurityType.Future, or SecurityType.Index is expected but was {underlyingSymbol.SecurityType}");
             }
 
+            // Resolve any mapping before requesting option contract list
```
Comment (Martin-Molinero) [inline Engine/DataFeeds/BacktestingOptionChainProvider.cs:59]: Maybe add a _why_ comment here 👍

**Review by Martin-Molinero — COMMENTED** [review]
Cool! Minor comment, let's assert behavior for indexes/futures just in case too

File: Engine/DataFeeds/BacktestingOptionChainProvider.cs (hunk 1 above)
Comment (gsalaz98) [reply Engine/DataFeeds/BacktestingOptionChainProvider.cs:61]: Adding on my review to this particular section of code, unrelated to MM's comment:

Are we sure we want to map futures/index options? 

The OptionChainProvider implementations were adjusted to generically support any kind of option chain being requested from it, regardless of SecurityType. 

Maybe we could have (or already have) some sort of extension method that checks if a Symbol or SecurityType requires mapping? 

Not a change that has to be added, but I do want to make sure we don't map Symbols for assets that don't need to be.

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**